pkg/model/diamond.go

```diff
@@ -0,0 +1,265 @@
+)
+
+// ConflictMode indicates the conflict detection mode defined for a diamond
+type ConflictMode string
+
+const (
+	// IgnoreConflicts is the diamond mode with which conflicts are not handled (latest win, no track kept of clobbered files)
+	IgnoreConflicts ConflictMode = "ignored"
+
+	// EnableCheckpoints is the diamond mode with which conflicts are explicitly handled and saved as "checkpoints" (incremental upload)
+	EnableCheckpoints ConflictMode = "enable-checkpoints"
+
+	// EnableConflicts is the diamond mode with which conflicts are detected and saved as "conflicts"
+	EnableConflicts ConflictMode = "enable-conflicts"
+
+	// ForbidConflicts is the diamond mode with which conflicts result in failure to commit
+	ForbidConflicts ConflictMode = "forbids-conflicts"
+)
+
+// DiamondState models the running status of an ungoing diamond workflow
+type DiamondState string
+
+const (
+	// DiamondInitialized is the state of an initialized diamond
+	DiamondInitialized DiamondState = "initialized"
+
+	// DiamondCommitting indicates that a commit is going on for this diamond
+	DiamondCommitting DiamondState = "committing"
+
+	// DiamondCancelling indicates that a cancel is going on for this diamond
+	DiamondCancelling DiamondState = "cancelling"
+
+	// DiamondRunning indicates that the diamond has started some splits
+	DiamondRunning DiamondState = "running"
+
+	// DiamondDone indicates the diamond has completed with a commit
+	DiamondDone DiamondState = "done"
+
+	// DiamondCancelled indicates the diamond has completed with a cancel
+	DiamondCancelled DiamondState = "cancelled"
+)
+
+const (
+	diamondDescriptorFile = "diamond.yaml"
+	splitDescriptorFile   = "split.yaml"
+)
+
+// IsValid checks the value of a diamond state
+func (s DiamondState) IsValid() bool {
+	switch s {
+	case DiamondInitialized, DiamondRunning, DiamondCommitting, DiamondDone, DiamondCancelling, DiamondCancelled:
+		return true
+	default:
+		return false
+	}
+}
+
+func (s DiamondState) String() string {
+	return string(s)
+}
+
+// SplitState models the running status of an ungoing diamond split
+type SplitState string
+
+const (
+	// SplitDone is the state of a completed split
+	SplitDone SplitState = "done"
+
+	// SplitFailed is the state of a failed split
+	SplitFailed SplitState = "failed"
+
+	// SplitRunning is the state of running split
+	SplitRunning SplitState = "running"
+)
+
+// IsValid checks the value of a diamond state
+func (s SplitState) IsValid() bool {
+	switch s {
+	case SplitDone, SplitFailed, SplitRunning:
+		return true
+	default:
+		return false
+	}
+}
+
+func (s SplitState) String() string {
+	return string(s)
+}
+
+// DiamondDescriptor models a diamond's metadata
+type DiamondDescriptor struct {
+	DiamondID       string            `json:"diamondID" yaml:"diamondID"`
+	StartTime       time.Time         `json:"startTime" yaml:"startTime"` // when the diamond was initialized
+	EndTime         time.Time         `json:"endTime,omitempty" yaml:"endTime,omitempty"`
```

Comment:
Having startime and endtime in the same model? This is probably introducing complicated tracking of metadata RMW to be correct.

Reply:
StartTime is set at initialization time, end time is updated when the diamond state transitions to a terminal state.

Reply:
i think RMW means "Read Mutate Write", and although i didn't plan on getting into whether to use such a RMW mechanism as part of the review of this initial sequence of commits, i did make [a comment](https://github.com/oneconcern/datamon/pull/403#discussion_r379651519) on an alternative (Write Once Read Many -type) mechanism